stage #0/module #4 Algorithms, Data Structures/algorithms_and_data_structures.md

```diff
@@ -0,0 +1,487 @@
+
+A data structure may be selected or designed to store data for the purpose of using it with various algorithms. It is not only important to use data structures, but it is also important to choose the proper data structure for each task.
+
+Data structures are divided into two categories:
+- Linear data structure (the elements are arranged in sequence one after the other, for example, arrays, linked lists, etc.);
+
+![Linear Data Structure](img/linear_data_structure.png)
+
+- Non-linear data structure (the elements are arranged in a hierarchical manner where one element will be connected to one or more elements, for example, hash table, tree, etc.).
+
+[comment]: <> (TODO copywriting)
+![Non-linear](img/non_linear_data_structure.png)
+
+### Arrays
+#### Overview 
+An array is a linear data structure:
+1) which occupies contiguous area of memory:
+
+![Contiguous Area of Memory](img/array_0.png)
+
+- a size of an array is defined while its creation;
+
+2) which consists of equal-type elements:
+
+![Equal-type Elements](img/array_1.png)
+
+- all elements of an array have the same type;
+
+3) which elements are indexed by contiguous integers:
+
+![Indexed Elements](img/array_2.png)
+
+- every element of an array has its own index. Each element can be accessed by index;
+- The first element of an array has a "zero" index in Java.  
+
+4) which has a fixed size:
+
+![Fixed Size](img/array_3.png)
+
+- a size of an array is defined at array's creation and can't be changed further;
+- in order to increase a size of an array, a new array with a desired size should be created, and it is necessary to copy elements there.
+
+![A new array with increased size](img/array_4.png)
+![A new array with increased size](img/array_5.png)
+
+#### Need of Arrays
+In computer programming, the most of the cases requires to store the large number of data of similar type. To store such amount of data, we need to define a large number of variables. It would be very difficult to remember names of all the variables while writing the programs. Instead of naming all the variables with a different name, it is better to define an array and store all the elements into it.
+
+### Arrays in Java
+#### Definition
+Arrays are defined and used with the square-brackets _indexing operator_ "**[]**".
+In order to define an array reference, 
+1) we should just write a desired type with empty square brackets:
+```
+int[] arrayOfIntegers; // declares an array of integers
+```
+2) we can also place the square-brackets after the identifier:
+```
+int arrayOfIntegers[]; // declares an array of integers
+```
+
+The above two definitions of an arrays have exactly the same meaning.
+But defining an array doesn't mean that we created an array, when array creation means allocating space in memory for storing elements. In order to create an array, we must write an initialization expression.
+
+#### Creation and Initialization
```

Comment:
maybe speaking in term of declaration and initialization is a better approach as a student will be getting used to it